### DISCUSSION TITLE
Is there a way to pass variable between different language code cells (from or to bash)?

### DISCUSSION THREAD

**Discussion by aborruso** [quarto-dev/quarto-cli#3701] [Q&A]
Hi,
is there a way to do something like this?

~~~~
```{bash}
var="hello"
```

```{python}
print(var)
```
~~~~

**Comment by mcanouil** [comment]
Languages are not kernel.
You can only use one kernel in a document, i.e., knitr, jypter, etc.

You need to use the proper syntax of the language in which you want to retrieve the information and/or make sure to export properly what you want to use.
For instance, look at how data are passed to Observable in the documentation: <https://quarto.org/docs/computations/ojs.html>.

Environment variables can usually be used by any languages, but you can store anything in it as long as it is a character string.

**Reply by aborruso** [reply]
> Languages are not kernel.

You are right. I have corrected my wrong subject.

**Reply by aborruso** [reply]
> For instance, look at how data are passed to Observable in the documentation: https://quarto.org/docs/computations/ojs.html.

Yes, it would be great to have a sort of general ojs_define function, to use it in all languages.

My goal is to create a variable in bash, and use it in another language cell, or simply in a markdown text.

> Environment variables can usually be used by any languages, but you can store anything in it as long as it is a character string.

Using this

~~~~
```{bash}
export myvar="hello"
```

I have written {{< env myvar >}}.

```{bash}
echo "$myvar world"
```
~~~~

I'm not able to get the variable value both in markdown and in another bash cell.

@mcanouil  Do you know how to get it?

Thank you

![image](https://user-images.githubusercontent.com/30607/208299145-8a8f1eed-1a5e-4b0b-af50-00d1ab698ddd.png)

**Reply by mcanouil** [reply]
I don't know for bash how it is handled in the different kernel/engine.
From R, it is pretty straightforward:

````md
---
title: "Untitled"
format: html
---

```{r}
Sys.setenv(PRODUCT_VERSION="1.1.1")
```

Version {{< env PRODUCT_VERSION >}} is a minor upgrade.

```{r}
Sys.getenv("PRODUCT_VERSION")
```
````

From Python:

````
---
title: "Untitled"
format: html
jupyter: python3
---

```{python}
import os
os.environ["PRODUCT_VERSION"] = "1.1.1"
```

Version {{< env PRODUCT_VERSION >}} is a minor upgrade.

```{python}
print(os.environ["PRODUCT_VERSION"])
```
````

From Python to R (via `knitr`):

````md
---
title: "Untitled"
format: html
---

```{python}
import os
os.environ["PRODUCT_VERSION"] = "1.1.1"
```

Version {{< env PRODUCT_VERSION >}} is a minor upgrade.

```{r}
Sys.getenv("PRODUCT_VERSION")
```
````

Note that, the env shortcode might be evaluated too early to get access to what has been generated in code cells (just a theory/guess, since I do not know the exact orders of all the steps).

**Reply by aborruso** [reply]
I know that it works in R and python, thank you very much.

But I need it in bash.

Probably there is not a bash direct way to do it :(

**Reply by mcanouil** [reply]
See why you can't at least with `knitr` as the engine/kernel: <https://stackoverflow.com/questions/60569395/set-environment-variable-in-bash-in-rmarkdown>

It's all about the order in which everything happen.
R/Jupyter are started before your bash code is evaluated, so it will remain unknown to R/Jupyter.

**Reply by aborruso** [reply]
There are great data things to do using simply the cli, and there are really impressive tools as awk, miller, qsv, jq, ecc. that for some tasks are IMHO more efficient than Python and R.

It seems to me a pity that this also works well everywhere except in bash.

Thank you @mcanouil

**Reply by mcanouil** [reply]
This is not something specific to Quarto/knitr/Jupyter.
This is basic behaviour.
If you open two terminals, they won't know what is happening within the other.
For example, I create a variable in one "session", the other "session" can't see it.
<img width="1460" alt="image" src="https://user-images.githubusercontent.com/8896044/208301958-38e55426-6ad6-4297-8f66-6bab9aecec59.png">

**Reply by aborruso** [reply]
> This is not something specific to Quarto/knitr/Jupyter.
> This is basic behaviour.

I don't want to say anything special. Just to reiterate that in the cli you do exceptional things with data, and it's a pity you can't pass variable values, a bit like what was done with observable cell and dedicated ojs_define function.

**Reply by mcanouil** [reply]
> cli you do exceptional things with data

As illustrated, this is only possible within the same session.

For `ojs_define` is only for R/Python at the moment.
There are too many languages, too many object structure to be handle by Quarto which is a language agnostic tool.
This being said, nothing forbids anyone to create/extend the reach of `ojs_define` or do similar job as `reticulate` and `rpy2` to only quote those.
See <https://github.com/quarto-dev/quarto-cli/blob/main/src/resources/rmd/ojs.R> and <https://github.com/quarto-dev/quarto-cli/blob/main/src/resources/rmd/ojs_static.R> for R.

**Reply by aborruso** [reply]
> As illustrated, this is only possible within the same session.

I swear I get it.

> For `ojs_define` is only for R/Python at the moment. There are too many languages, too many object structure to be handle by Quarto which is a language agnostic tool. This being said, nothing forbids anyone to create/extend the reach of `ojs_define` or do similar job as `reticulate` and `rpy2` to only quote those.

Unfortunately I am not able to do it, otherwise I would write it myself. I took the liberty of repeating myself, only because I think that building bridges to bash is an important thing, because you can do a lot of great data things there.

Thank you @mcanouil

**Reply by mcanouil** [reply]
I agree that extending to bash (also PowerShell for Windows users) could be great, but not sure it's going to happen at least in a near future for `ojs_define`.

Note that contrary to `ojs_define` defined by Quarto, for the bash thing using `knitr` as engine, the best target for a feature request would be the `knitr` repository and when using `jupyter`, well the official repository for it.
Quarto only uses those engine/kernels. 😉

**Reply by cscheid** [reply]
(I've edited your posts to remove typos, just FYI)

**Reply by aborruso** [reply]
@mcanouil @cscheid in some way I found what I was looking for, using `system` in r.

One of my goals is sometimes to use a bash data analysis utility and collect the output in a variable and pass it to r, to use the convenient syntax ``The sum of this is `r sum` ``.

In the below example I use [Miller](https://miller.readthedocs.io/en/latest/) to make a sum. I get the result and I use it in markdown.

And using `reticulate` I can pass it to python.

I'm sure you don't like it and it's not elegant, but since I can't do it from a bash block, I've been looking for something like python magic in a r block.

A stupid question: how can I not put the absolute path of the CSV file, and insert a variable that prints the current path (the qmd and the CSV are in the same folder)?

Best regards

````
```{r}
sum <- system('mlr --c2n stats1 -a sum -f a /home/andy/quarto_test/input.csv', intern = TRUE)
```

The sum is `r sum`.

```{python}
sumpy = r.sum
print(sumpy)
```
````

**Reply by cscheid** [reply]
Something like this?

````
```{r}
cmd <- paste0(c("mlr --c2n stats1 -a sum -f a ", getwd(), "input.csv"), collapse="")
sum <- system(cmd, intern = TRUE)
```
````

But I'm a bit confused, because I think that simply omitting the path should work; the working directory for the system() call should be that of the document.
